### PR TITLE
Add provide-clock property to omxcamera element to be able of sync audio against video stream

### PR DESCRIPTION
**Pull request by cgomezv**
Add provide-clock property to omxcamera element to be able of sync audio against video stream

### CODE REVIEW

File: omx/gstomxcamera.c [hunk 1]
```diff
@@ -1071,3 +1125,49 @@ gst_omx_camera_create (GstPushSrc * src, GstBuffer ** buf)
...
+
+  GST_OBJECT_LOCK (self);
+  if (!self->provide_clock)
+    goto clock_disabled;
+
+  clock = GST_CLOCK_CAST (gst_object_ref (self->clock));
+
+  GST_OBJECT_UNLOCK (self);
+
+
```
Comment (joselopez10) [inline omx/gstomxcamera.c]: If not needed, remove this empty line

File: omx/gstomxcamera.c [hunk 2]
```diff
@@ -268,6 +285,15 @@ gst_omx_camera_init (GstOMXCamera * self)
...
   self->skip_frames = PROP_SKIP_FRAMES_DEFAULT;
+  self->provide_clock = PROP_PROVIDE_CLOCK_DEFAULT;
+  if (PROP_PROVIDE_CLOCK_DEFAULT)
+    GST_OBJECT_FLAG_SET (self, GST_ELEMENT_FLAG_PROVIDE_CLOCK);
+  else
+    GST_OBJECT_FLAG_UNSET (self, GST_ELEMENT_FLAG_PROVIDE_CLOCK);
+
+  self->clock =
+      g_object_new (GST_TYPE_OMX_CLOCK, "name", "GstOmxCameraClock", NULL);
+  g_object_set (self->clock, "clock-type", GST_CLOCK_TYPE_OTHER, NULL);
```
Comment (michaelgruner) [inline omx/gstomxcamera.c]: This should be in the Clock class not here

File: omx/gstomxcamera.c [hunk 3]
```diff
@@ -341,6 +367,15 @@ gst_omx_camera_set_property (GObject * object,
       if (self->comp)
         gst_omx_camera_set_skip_frames (self);
       break;
+    case PROP_PROVIDE_CLOCK:
+      self->provide_clock = g_value_get_boolean (value);
```
Comment (michaelgruner) [inline omx/gstomxcamera.c]: Move this inside the mutex lock

File: omx/gstomxcamera.c [hunk 4]
```diff
@@ -392,6 +427,9 @@ gst_omx_camera_get_property (GObject * object,
       }
       break;
     }
+    case PROP_PROVIDE_CLOCK:
+      g_value_set_boolean (value, self->provide_clock);
```
Comment (michaelgruner) [inline omx/gstomxcamera.c]: Serialize this as well

File: omx/gstomxcamera.c [hunk 5]
```diff
@@ -1071,3 +1125,49 @@ gst_omx_camera_create (GstPushSrc * src, GstBuffer ** buf)
...
+  GST_OBJECT_LOCK (self);
+  if (!self->provide_clock)
+    goto clock_disabled;
+
+  clock = GST_CLOCK_CAST (gst_object_ref (self->clock));
+
+  GST_OBJECT_UNLOCK (self);
+
+
+  return clock;
```
Comment (michaelgruner) [inline omx/gstomxcamera.c]: In the case of `clock_disabled` you are returning an uninitialized `GstClock *`. You should initialize `clock` to NULL. This should've given you a warning.

File: omx/gstomxclock.c [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+
+  monotonic = g_get_monotonic_time ();
+  now.tv_usec = monotonic % G_USEC_PER_SEC;
+  now.tv_sec = monotonic / G_USEC_PER_SEC;
+
+  return GST_TIMEVAL_TO_TIME (now);
+}
+
+static GstClockTime
+gst_omx_clock_get_internal_time (GstClock * clock)
```
Comment (michaelgruner) [inline omx/gstomxclock.c]: pointer guard

File: omx/gstomxclock.c [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+    time = self->last_tick + offset + self->hw_offset;
+    self->last_time_sent = time;
+  }
+
+  GST_DEBUG_OBJECT (self, "time %" GST_TIME_FORMAT, GST_TIME_ARGS (time));
+  return time;
+}
+
+void
+gst_omx_clock_new_tick (GstOmxClock * self, GstClockTime tick)
```
Comment (michaelgruner) [inline omx/gstomxclock.c]: Pointer guard

File: omx/gstomxclock.c [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+
+  GST_DEBUG_OBJECT (self,
+      "last tick %" GST_TIME_FORMAT ", last time %" GST_TIME_FORMAT,
+      GST_TIME_ARGS (self->last_tick), GST_TIME_ARGS (self->last_time));
+
+  GST_OBJECT_UNLOCK (self);
+}
+
+void
+gst_omx_clock_reset (GstOmxClock * self)
```
Comment (michaelgruner) [inline omx/gstomxclock.c:174]: Pointer guard

**Review by michaelgruner — CHANGES_REQUESTED** [review]
Please cleanup build warnings.

File: omx/gstomxcamera.c (hunk 2 above)
Comment (cgomezv) [reply omx/gstomxcamera.c]: Ok, moved to gst_omx_clock_init

File: omx/gstomxcamera.c (hunk 5 above)
Comment (cgomezv) [reply omx/gstomxcamera.c]: You are right, already fixed!

**Review by michaelgruner — APPROVED** [review]

**Pull request merged**